### ISSUE TITLE
Link Error in ubuntu 14.04 and 16.04

### ISSUE THREAD

**Issue by TwIStOy** [MaskRay/ccls#111]
I try to build ccls on ubuntu 14.04 and ubuntu 16.04, but got a lot of link error.
```
cmake .. -DSYSTEM_CLANG=OFF -DCMAKE_CXX_COMPILER=/root/ccls/Release/clang+llvm-7.0.0-x86_64-linux-gnu-ubuntu-14.04/bin/clang++ -DCMAKE_PREFIX_PATH=/root/ccls/Release/clang+llvm-7.0.0-x86_64-linux-gnu-ubuntu-14.04 -DCLANG_USE_BUNDLED_LIBC++=ON -DCMAKE_PREFIX_PATH=/root/env -DCMAKE_CXX_LINK_FLAGS=-fuse-ld=lld
```
Some of errors:
```
ld.lld: error: undefined symbol: clang::PrecompiledPreamble::Build(clang::CompilerInvocation const&, llvm::MemoryBuffer const*, clang::PreambleBounds, clang::DiagnosticsEngine&, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>, std::__1::shared_ptr<clang::PCHContainerOperations>, bool, clang::PreambleCallbacks&)
>>> referenced by clang_complete.cc
>>>               CMakeFiles/ccls.dir/src/clang_complete.cc.o:(ccls::(anonymous namespace)::CompletionPreloadMain(void*))

ld.lld: error: undefined symbol: clang::CompilerInstance::CompilerInstance(std::__1::shared_ptr<clang::PCHContainerOperations>, clang::MemoryBufferCache*)
>>> referenced by clang_complete.cc
>>>               CMakeFiles/ccls.dir/src/clang_complete.cc.o:(ccls::(anonymous namespace)::BuildCompilerInstance(ccls::CompletionSession&, std::__1::unique_ptr<clang::CompilerInvocation, std::__1::default_delete<clang::CompilerInvocation> >, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>, clang::DiagnosticConsumer&, ccls::PreambleData const*, ccls::WorkingFiles::Snapshot const&, std::__1::vector<std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> >, std::__1::allocator<std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> > > >&))

ld.lld: error: undefined symbol: clang::CompilerInstance::setInvocation(std::__1::shared_ptr<clang::CompilerInvocation>)
>>> referenced by clang_complete.cc
>>>               CMakeFiles/ccls.dir/src/clang_complete.cc.o:(ccls::(anonymous namespace)::BuildCompilerInstance(ccls::CompletionSession&, std::__1::unique_ptr<clang::CompilerInvocation, std::__1::default_delete<clang::CompilerInvocation> >, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>, clang::DiagnosticConsumer&, ccls::PreambleData const*, ccls::WorkingFiles::Snapshot const&, std::__1::vector<std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> >, std::__1::allocator<std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> > > >&))

ld.lld: error: undefined symbol: clang::TargetInfo::CreateTargetInfo(clang::DiagnosticsEngine&, std::__1::shared_ptr<clang::TargetOptions> const&)
>>> referenced by clang_complete.cc
>>>               CMakeFiles/ccls.dir/src/clang_complete.cc.o:(ccls::(anonymous namespace)::BuildCompilerInstance(ccls::CompletionSession&, std::__1::unique_ptr<clang::CompilerInvocation, std::__1::default_delete<clang::CompilerInvocation> >, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>, clang::DiagnosticConsumer&, ccls::PreambleData const*, ccls::WorkingFiles::Snapshot const&, std::__1::vector<std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> >, std::__1::allocator<std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> > > >&))

ld.lld: error: undefined symbol: llvm::cl::SetVersionPrinter(std::__1::function<void (llvm::raw_ostream&)>)
```

**Comment by MaskRay** [comment]
See https://github.com/MaskRay/ccls/wiki/Build#cmake-options

`cmake -GNinja -H. -BBundled -DSYSTEM_CLANG=on -DCMAKE_PREFIX_PATH=$PWD/build/clang+llvm-7.0.0-x86_64-linux-gnu-ubuntu-14.04`

The prebuilt archive for Ubuntu 14.04 links against libstdc++. You shall not use `-DCLANG_USE_BUNDLED_LIBC++=ON`

**Comment by TwIStOy** [comment]
It doesn't work.
```
/home/haotian/ccls/src/maybe.hh:18:10: fatal error: 'optional' file not found
#include <optional>
         ^~~~~~~~~~
1 error generated.
```

**Comment by TwIStOy** [comment]
```
cmake .. \
  -GNinja \
  -DSYSTEM_CLANG=ON \
  -DCMAKE_CXX_COMPILER=/home/haotian/ccls/Release/clang+llvm-7.0.0-x86_64-linux-gnu-ubuntu-14.04/bin/clang++ \
  -DCMAKE_PREFIX_PATH=/home/haotian/ccls/Release/clang+llvm-7.0.0-x86_64-linux-gnu-ubuntu-14.04 \
  -DCMAKE_CXX_LINK_FLAGS=-fuse-ld=lld
```

Got

```
CMake Error at CMakeLists.txt:18 (add_executable):
  Target "ccls" links to target "Clang::Clang" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
```

**Comment by MaskRay** [comment]
https://github.com/MaskRay/ccls/wiki/Getting-started#build-the-ccls-language-server

On Linux, building requires libstdc++ shipped with GCC >= 7.3 for C++17 header files, even if you use clang++.

**Comment by MaskRay** [comment]
This issue report was impolite. You didn't give the compiler version. Closing
